### ISSUE TITLE
SD card image not working

### ISSUE THREAD

**Issue by ste7anste7an** [rick-heig/eNVMe#1]
I downloaded the eNVMe v1.0.0 release sd card image file [sdcard_envme_ubuntu2404_241106.img.xz](https://github.com/rick-heig/eNVMe/releases/download/v1.0.0/sdcard_envme_ubuntu2404_241106.img.xz) and copied it to a 16GByte sd card using

```xzcat sdcard_envme_ubuntu2404_241106.img.xz | sudo dd of=/dev/<your sd card> bs=1M && sudo sync```

I use a nanoPC-T6 (16GB/64GB) version. I could not boot from the sd card. 

I tried the image `rk3588-sd-ubuntu-noble-desktop-6.1-arm64-20241112.img.gz` from the offical FriendlyElec repository (google drive), flashed that and that works.

When I do `fdisk -l` on the envme image, it just finds a single partition:
```
Disk envme_sdcard.img: 7.01 GiB, 7526678528 bytes, 14700544 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x00000000

Device            Boot Start      End  Sectors Size Id Type
envme_sdcard.img1          1 14700543 14700543   7G ee GPT
```
On the official image I find 8 partition among which two ext4 partitions.

COuld you pelase check the validity of the sd card image?

**Comment by rick-heig** [comment]
Hello,
The official image and the one provided are very different that is normal, the official image is a 6.1 Linux kernel with many things patched for that board by rockchip. I provide a minimal image with Ubuntu server without graphics and a mainline 6.12 kernel.

The 7GB size of the partition is correct, is was chosen to fit 8GB cards.

How did you check the boot ? What does the UART serial console give ?

If you just checked HDMI output it is normal you won't get anything as I did not create the image to use video output, because at the time video output support for that board was not yet integrated in the mainline linux kernel.

So the image should boot and you can get the serial console over UART, then later you can use SSH alternatively.

Let me know what you get over UART. 

I should clarify that the provided image is without graphical support, I cannot use the "official" friendlyelec kernel as it is version 6.1 and lacks necessary features for eNVMe and the mainline linux kernels dont have full HDMI support yet for the RK3588.

You can follow the development status here https://gitlab.collabora.com/hardware-enablement/rockchip-3588/notes-for-rockchip-3588/-/blob/main/mainline-status.md

Best regards,
Rick

**Comment by ste7anste7an** [comment]
[log_serial_boot.txt](https://github.com/user-attachments/files/17781413/log_serial_boot.txt)

This is the logging I get from the serial port when booting the envme sd card image. I checked the sha256 for the xz file, and that was correct.

It seems that it can't find a bootable partition onthe sd card, and then it decides to boot from internal flash (that's android).

What should be the sha256 of the uncompressed image? What do you get when you do fdisk -l on the uncompressed image?

**Comment by rick-heig** [comment]
I will have a look on Monday as I don't have access to the hardware over the weekend.

The log seems to be from the boot from the official image either from SD or eMMC (as it shows an older version of U-Boot compared to the one I use)

The inage I made has more recent U-boot and Linux, and from the log the U-boot doesn't seem to be the one from my Image.

If you want to rebuild an image from scratch you can follow the instructions here : https://github.com/rick-heig/eNVMe/blob/main/doc/platform.md just do steps 1 and 3, they should be enough to get a bootable image (with minimal rootfs). The image after step 3 can be tested directly, even if it is for the CM3588, it should at least boot into U-Boot and normally the kernel as well. You can check the partition table of that image directly with fdisk to compare.

Do you have the LTS version of the T6 ? I only have the older T6 for the moment, but ordered the LTS version this week, so that I can make sure it works the same on that board as well. However I dont think there should be a difference between the two boards. But maybe the boot has some peculiarities.

I'll check more on Monday.
Have a nice weekend.
Rick

**Comment by ste7anste7an** [comment]
We succeeded in making our own bootable SD and copied the rootfs from your image over to that one. It boots from that sd card and the nvme endpoint is wokring (sort of). The links keeps toggling up and down and for short periods of time it works. We will try decreasing the pcie speed to pcie2.

So, if you would like to look next week at your sd card image, that would be great, as I have the impression that the GPT and uboot part is somehow corrupted.

BTW How can I see which version NanoPC-T6 I have (LTS or not)?

**Comment by rwk-unil** [comment]
Hello,
Thank you for the update.

For the link going up and down, this is because the endpoint works in separate clock mode and I can't do anything to improve this, my observations were that is was very dependent on the host PC and cabling (some work fine, others have much more link instabilities), and as described here https://github.com/rick-heig/eNVMe?tab=readme-ov-file#known-issues you can try the overlay for PCIe 2.0 mode.

I have tried to SD card image again and I was able to boot directly :
```
xzcat sdcard_envme_ubuntu2404_241106.img.xz | sudo dd of=/dev/sdg bs=1M && sudo sync
```

This would boot U-boot and the kernel

(However, linux boot got stuck because I didn't change `/boot/extlinux/extlinux.conf` to take the T6 device trees, once changing this, it boots fully into the Ubuntu 24.04).

Boot log looks like this : 
```
DDR V1.10 75d050770f typ 23/02/28-20:47:10
LPDDR4X, 2112MHz
channel[0] BW=16 Col=10 Bk=8 CS0 Row=17 CS1 Row=17 CS=2 Die BW=8 Size=4096MB
channel[1] BW=16 Col=10 Bk=8 CS0 Row=17 CS1 Row=17 CS=2 Die BW=8 Size=4096MB
channel[2] BW=16 Col=10 Bk=8 CS0 Row=17 CS1 Row=17 CS=2 Die BW=8 Size=4096MB
channel[3] BW=16 Col=10 Bk=8 CS0 Row=17 CS1 Row=17 CS=2 Die BW=8 Size=4096MB
Manufacturer ID:0x1
CH0 RX Vref:27.9%, TX Vref:22.8%,22.8%
CH1 RX Vref:26.7%, TX Vref:23.8%,22.8%
CH2 RX Vref:26.3%, TX Vref:22.8%,21.8%
CH3 RX Vref:27.1%, TX Vref:23.8%,22.8%
change to F1: 528MHz
change to F2: 1068MHz
change to F3: 1560MHz
change to F0: 2112MHz
out
U-Boot SPL board init
U-Boot SPL 2017.09-gd91b448214-221107 #root (Apr 10 2023 - 14:05:21)
Trying to boot from MMC2
spl: partition error
Trying fit image at 0x4000 sector
## Verified-boot: 0
## Checking atf-1 0x00040000 ... sha256(b40ce543bb...) + OK
## Checking u-boot 0x00a00000 ... sha256(893641a38c...) + OK
## Checking fdt-1 0x00ae03a0 ... sha256(3e56afa830...) + OK
## Checking atf-2 0x000f0000 ... sha256(f202de33d9...) + OK
## Checking atf-3 0xff100000 ... sha256(af08976afd...) + OK
Jumping to U-Boot(0x00a00000) via ARM Trusted Firmware(0x00040000)
Total: 763.984 ms

INFO:    Preloader serial: 2
NOTICE:  BL31: v2.3():v2.3-616-gdc1125f48:finley.xiao
NOTICE:  BL31: Built : 16:41:17, Jul 13 2023
INFO:    spec: 0x1
INFO:    ext 32k is not valid
INFO:    ddr: stride-en 4CH
INFO:    GICv3 without legacy support detected.
INFO:    ARM GICv3 driver initialized in EL3
INFO:    valid_cpu_msk=0xff bcore0_rst = 0x0, bcore1_rst = 0x0
INFO:    system boots from cpu-hwid-0
INFO:    idle_st=0x21fff, pd_st=0x11fff9, repair_st=0xfff70001
ERROR:   dfs get fsp_params[0] error, 0xfead0001 != 0xfead0003
ERROR:   dfs get fsp_params[1] error, 0xa4c != 0xfead0003
ERROR:   dfs get fsp_params[2] error, 0xa44 != 0xfead0003
ERROR:   dfs get fsp_params[3] error, 0x4000001 != 0xfead0003
ERROR:   loader&trust unmatch!!! Please update loader if need enable dmc
INFO:    BL31: Initialising Exception Handling Framework
INFO:    BL31: Initializing runtime services
WARNING: No OPTEE provided by BL2 boot loader, Booting device without OPTEE initialization. SMC`s destined for OPTEE will return SMC_UNK
ERROR:   Error initializing runtime service opteed_fast
INFO:    BL31: Preparing for EL3 exit to normal world
INFO:    Entry point address = 0xa00000
INFO:    SPSR = 0x3c9


U-Boot 2024.07 (Oct 25 2024 - 15:13:10 +0200)

Model: Radxa ROCK 5 Model B
DRAM:  16 GiB
Core:  353 devices, 32 uclasses, devicetree: separate
MMC:   mmc@fe2c0000: 1, mmc@fe2d0000: 2, mmc@fe2e0000: 0
Loading Environment from nowhere... OK
In:    serial@feb50000
Out:   serial@feb50000
Err:   serial@feb50000
Model: Radxa ROCK 5 Model B
Net:   No ethernet found.
Hit any key to stop autoboot:  0 
Scanning for bootflows in all bootdevs
Seq  Method       State   Uclass    Part  Name                      Filename
---  -----------  ------  --------  ----  ------------------------  ----------------
Scanning global bootmeth 'efi_mgr':
Card did not respond to voltage select! : -110
No EFI system partition
No EFI system partition
Failed to persist EFI variables
No EFI system partition
Failed to persist EFI variables
No EFI system partition
Failed to persist EFI variables
No EFI system partition
Failed to persist EFI variables
No EFI system partition
Failed to persist EFI variables
  0  efi_mgr      ready   (none)       0  <NULL>                    
** Booting bootflow '<NULL>' with efi_mgr
Loading Boot0000 'mmc 1' failed
Loading Boot0001 'mmc 0' failed
EFI boot manager: Cannot load any image
Boot failed (err=-14)
Scanning bootdev 'mmc@fe2c0000.bootdev':
  1  extlinux     ready   mmc          1  mmc@fe2c0000.bootdev.part /boot/extlinux/extlinux.conf
** Booting bootflow 'mmc@fe2c0000.bootdev.part_1' with extlinux
1:	FriendlyElec CM3588 NAS Kit Linux
Retrieving file: /boot/Image
append: root=/dev/mmcblk1p1 ro earlycon rootwait
Retrieving file: /boot/rk3588-nanopc-t6.dtb
Retrieving file: /boot/rk3588-nanopc-t6-pcie-ep.dtbo
## Flattened Device Tree blob at 12000000
   Booting using the fdt blob at 0x12000000
Working FDT set to 12000000
   Loading Device Tree to 00000000ecb76000, end 00000000ecba3fff ... OK
Working FDT set to ecb76000

Starting kernel ...

[    0.000000] Booting Linux on physical CPU 0x0000000000 [0x412fd050]
[    0.000000] Linux version 6.12.0-rc3 (rick@A13PC04R) (aarch64-none-linux-gnu-gcc (Arm GNU Toolchain 13.2.rel1 (Build arm-13.7)) 13.2.1 20231009, GNU ld (Arm GNU Toolchain 13.2.rel1 (Build arm-13.7)) 2.41.0.20231009) #12 SMP Wed Nov  6 16:41:45 CET 2024
...

Ubuntu 24.04.1 LTS ENVME ttyS2

ENVME login: 
```

Here are the sha256sum values for both the compressed and uncompressed images :

```
2e422a9bc1262b209f8e1a2966f709f6793bccd88f3649b6d9a188cbf7d13fbc  sdcard_envme_ubuntu2404_241106.img.xz
7ced4f7e4dc5320207bb21d5cf9d82a3185e50b2b956a0e825001517191924da  sdcard_envme_ubuntu2404_241106.img
```

You can check the difference between the two T6 boards here : https://wiki.friendlyelec.com/wiki/index.php/NanoPC-T6#NanoPC-T6-LTS

I cannot reproduce your issue, so I can't do much more for the moment. I have ordered the LTS version of the T6 to check if there is a difference in the boot, I'll check when I get the board.

If you find out what is wrong, e.g., by comparing the working image you have now and the image I provided (e.g., partition table differences etc.) and find a solution feel free to share, I'll patch and update the image.

Best regards,
Rick

**Comment by ste7anste7an** [comment]
Hi,

I checked the hashes. Th .xz file is correct, but unpacking the xz using resulted in an incorrect checksum for the uncompressed file. Now I decompressed the file using `xz -d ` and now I do get the correct hash.

Will try to boot from the image later.

Thanks.

**Comment by ste7anste7an** [comment]
Flashed the image (with the correct hash) to an SD card. It still does not boot from the sdcard and starts booting from the internal eMMC.

An `fdisk -l /dev/sda` on our self-created image gives:
```
Disk /dev/sda: 29.72 GiB, 31914983424 bytes, 62333952 sectors
Disk model: Storage Device  
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: FC5DB271-CBCE-4B2E-80C5-8173CD62BAF6

Device     Start      End  Sectors  Size Type
/dev/sda1  18864 62333918 62315055 29.7G Linux filesystem
```

Note, that we expanded the image to the full 32GByte of the sd card

An `fdisk -l /dev/sda` on the sd card created based on the `sdcard_envme_ubuntu2404_241106.img` gives:
```
GPT PMBR size mismatch (62333951 != 15415295) will be corrected by write.
Disk /dev/sda: 7.35 GiB, 7892631552 bytes, 15415296 sectors
Disk model: Storage Device  
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x00000000

Device     Boot Start      End  Sectors  Size Id Type
/dev/sda1           1 15415295 15415295  7.4G ee GPT

```

Are you sure, your board boots from the SD card and not from the internal eMMC?

**Comment by rwk-unil** [comment]
Hello.

I have tested to boot without the SD card and I get the log shown below.
So the first code that executes seems to be the same in both cases (you have my other log above), also the U-Boot SPL seems to be the same, it differs when jumping to U-Boot.
The preloader version is different, and then the U-Boot and Kernel are different.

So you might be on something here, it might be that the first code that executes is on the eMMC and then in my case jumps to the U-Boot on the SD card while in your case with the flashed image it doesn't (but does when you create the image yourself).

Could you compare the boot log when you boot with the SD image you prepared yourself and the one you flashed ?

You could also compare the created image and flashed image with fdisk, I think the created image has an hybrid partition table (GPT+MBR), maybe something is wrong there in the image.

As I will receive new T6 boards soon, which will probably have the same eMMC contents as yours, I will compare with them, because with the two I currently have (older T6 2301, not the LTS version) the flashed image works.

I'll let you know when I have the received the new boards, for the moment I cannot reproduce the issue so it is difficult for me to fix this, especially since when you make the image by hand it works, and the image I made was made in the same way...

My eMMC (without SD) boot log :
```
DDR V1.10 75d050770f typ 23/02/28-20:47:10
LPDDR4X, 2112MHz
channel[0] BW=16 Col=10 Bk=8 CS0 Row=17 CS1 Row=17 CS=2 Die BW=8 Size=4096MB
channel[1] BW=16 Col=10 Bk=8 CS0 Row=17 CS1 Row=17 CS=2 Die BW=8 Size=4096MB
channel[2] BW=16 Col=10 Bk=8 CS0 Row=17 CS1 Row=17 CS=2 Die BW=8 Size=4096MB
channel[3] BW=16 Col=10 Bk=8 CS0 Row=17 CS1 Row=17 CS=2 Die BW=8 Size=4096MB
Manufacturer ID:0x1
CH0 RX Vref:27.9%, TX Vref:24.8%,23.8%
CH1 RX Vref:26.7%, TX Vref:23.8%,22.8%
CH2 RX Vref:26.3%, TX Vref:22.8%,21.8%
CH3 RX Vref:27.1%, TX Vref:23.8%,22.8%
change to F1: 528MHz
change to F2: 1068MHz
change to F3: 1560MHz
change to F0: 2112MHz
out
U-Boot SPL board init
U-Boot SPL 2017.09-gd91b448214-221107 #root (Apr 10 2023 - 14:05:21)
Trying to boot from MMC2
MMC: no card present
mmc_init: -123, time 0
spl: mmc init failed with error: -123
Trying to boot from MMC1
Trying fit image at 0x4000 sector
## Verified-boot: 0
## Checking atf-1 0x00040000 ... sha256(ac799e4aa0...) + OK
## Checking uboot 0x00200000 ... sha256(ecc2ca052f...) + OK
## Checking fdt 0x0032ce40 ... sha256(e936f08b25...) + OK
## Checking atf-2 0xff100000 ... sha256(c0f2f7769f...) + OK
## Checking atf-3 0x000f0000 ... sha256(338cd7cb83...) + OK
## Checking optee 0x08400000 ... sha256(fde0860845...) + OK
Jumping to U-Boot(0x00200000) via ARM Trusted Firmware(0x00040000)
Total: 517.222 ms

INFO:    Preloader serial: 2
NOTICE:  BL31: v2.3():v2.3-557-g9609b9c19:derrick.huang
NOTICE:  BL31: Built : 14:29:57, Mar  8 2023
INFO:    spec: 0x1
INFO:    ext 32k is not valid
INFO:    ddr: stride-en 4CH
INFO:    GICv3 without legacy support detected.
INFO:    ARM GICv3 driver initialized in EL3
INFO:    valid_cpu_msk=0xff bcore0_rst = 0x0, bcore1_rst = 0x0
INFO:    system boots from cpu-hwid-0
INFO:    idle_st=0x21fff, pd_st=0x11fff9, repair_st=0xfff70001
INFO:    dfs DDR fsp_params[0].freq_mhz= 2112MHz
INFO:    dfs DDR fsp_params[1].freq_mhz= 528MHz
INFO:    dfs DDR fsp_params[2].freq_mhz= 1068MHz
INFO:    dfs DDR fsp_params[3].freq_mhz= 1560MHz
INFO:    BL31: Initialising Exception Handling Framework
INFO:    BL31: Initializing runtime services
INFO:    BL31: Initializing BL32
I/TC: 
I/TC: OP-TEE version: 3.13.0-652-g4542e1efd #derrick.huang (gcc version 10.2.1 20201103 (GNU Toolchain for the A-profile Architecture 10.2-2020.11 (arm-10.16))) #5 2022年 09月 20日 星期二 09:41:09 CST aarch64
I/TC: Primary CPU initializing
I/TC: Primary CPU switching to normal world boot
INFO:    BL31: Preparing for EL3 exit to normal world
INFO:    Entry point address = 0x200000
INFO:    SPSR = 0x3c9


U-Boot 2017.09-gd91b448214-221107 #root (Apr 10 2023 - 14:05:21 +0800)

Board: NanoPi R6S
PreSerial: 2, raw, 0xfeb50000
DRAM:  16 GiB
Sysmem: init
Relocation Offset: eda4a000
Relocation fdt: eb9fab18 - eb9fecd8
CR: M/C/I
Using default environment

mmc@fe2c0000: 1, mmc@fe2e0000: 0
Bootdev(atags): mmc 0
MMC0: HS200, 200Mhz
PartType: EFI
DM: v2
boot mode: None
Found DTB in resource part
DTB: rk3588-nanopi6-rev01.dtb (251880)
HASH(c): OK
ANDROID: fdt overlay OK
I2c0 speed: 100000Hz
vsel-gpios- not found!
en-gpios- not found!
vdd_cpu_big0_s0 800000 uV
vsel-gpios- not found!
en-gpios- not found!
vdd_cpu_big1_s0 800000 uV
I2c2 speed: 100000Hz
vsel-gpios- not found!
en-gpios- not found!
vdd_npu_s0 800000 uV
spi2: RK806: 2
ON=0x00, OFF=0x00
vdd_gpu_s0 750000 uV
vdd_cpu_lit_s0 750000 uV
vdd_log_s0 750000 uV
vdd_vdenc_s0 init 750000 uV
vdd_ddr_s0 850000 uV
get vp0 plane mask:0x5, primary id:2, cursor_plane:0, from dts
get vp1 plane mask:0xa, primary id:3, cursor_plane:-1, from dts
get vp2 plane mask:0x140, primary id:8, cursor_plane:-1, from dts
Could not find baseparameter partition
I2c6 speed: 200000Hz
Rockchip UBOOT DRM driver version: v1.0.1
vp0 have layer nr:2[0 2 ], primary plane: 2
vp1 have layer nr:2[1 3 ], primary plane: 3
vp2 have layer nr:2[6 8 ], primary plane: 8
vp3 have layer nr:0[], primary plane: 0
hdmi@fde80000 disconnected
CLK: (uboot. arm: enter 1008000 KHz, init 1008000 KHz, kernel 0N/A)
  b0pll 24000 KHz
  b1pll 24000 KHz
  lpll 24000 KHz
  v0pll 24000 KHz
  aupll 786431 KHz
  cpll 1500000 KHz
  gpll 1188000 KHz
  npll 850000 KHz
  ppll 1100000 KHz
  aclk_center_root 702000 KHz
  pclk_center_root 100000 KHz
  hclk_center_root 396000 KHz
  aclk_center_low_root 500000 KHz
  aclk_top_root 594000 KHz
  pclk_top_root 100000 KHz
  aclk_low_top_root 396000 KHz
vdd_usbc 12220 mV

Net:   No ethernet found.
Hit key to stop autoboot('CTRL+C'):  0 
## Booting FIT Image FIT: No fit blob
FIT: No FIT image
ANDROID: reboot reason: "(none)"
optee api revision: 2.0
TEEC: Waring: Could not find security partition
Not AVB images, AVB skip
No valid android hdr
Android image load failed
Android boot failed, error -1.

## Booting Rockchip Format Image
fdt	 @ 0x08300000 (0x0003da88)
kernel   @ 0x00400000 (0x02107008)
ramdisk  @ 0x0a200000 (0x007b2bc0)
Fdt Ramdisk skip relocation
## Flattened Device Tree blob at 0x08300000
   Booting using the fdt blob at 0x08300000
  'reserved-memory' cma: addr=10000000 size=8000000
  'reserved-memory' ramoops@110000: addr=110000 size=e0000
   Using Device Tree in place at 0000000008300000, end 0000000008340a87
Adding bank: 0x00200000 - 0x08400000 (size: 0x08200000)
Adding bank: 0x09400000 - 0xf0000000 (size: 0xe6c00000)
Adding bank: 0x100000000 - 0x3fc000000 (size: 0x2fc000000)
Adding bank: 0x3fc500000 - 0x3fff00000 (size: 0x03a00000)
Adding bank: 0x4f0000000 - 0x500000000 (size: 0x10000000)
Total: 1590.728 ms

Starting kernel ...

[    2.616548] Booting Linux on physical CPU 0x0000000000 [0x412fd050]
[    2.616570] Linux version 5.10.110 (root@jensen) (aarch64-linux-gcc (ctng-1.25.0-119g-FA) 11.3.0, GNU ld (GNU Binutils) 2.38) #15 SMP Fri May 26 10:44:08 CST 2023
```

**Comment by ste7anste7an** [comment]
Note that I flashed the image from the Google drive under: 1_Official Images -> 2_SD-to-eMMC Images -> rk3588-eflasher-debian-bullseye-minimal-6.1-arm64-20241112.img.gz
That installs the debian on the eMMC card. We used that to prepare the new SD card image. Initially there was Android installed on the eMMC on the 16GB/64GB LTS model of NanoPC-T6.

**Comment by rwk-unil** [comment]
Hello,
I have received two FriendlyElec NanoPC-T6 LTS boards, I flashed the SD card with :

```
xzcat sdcard_envme_ubuntu2404_241106.img.xz | sudo dd of=/dev/<your sd card> bs=1M && sudo sync
```

I can confirm It boots directly into the right kernel (log below) same as with the older NanoPC-T6 2301 version.
I can't do much more as I cannot reproduce your issue. But at least you managed to make your own bootable image.

Best regards,
Rick

```
DDR 9fffbe1e78 cym 24/02/04-10:09:20,fwver: v1.16
LPDDR4X, 2112MHz
channel[0] BW=16 Col=10 Bk=8 CS0 Row=17 CS1 Row=17 CS=2 Die BW=8 Size=4096MB
channel[1] BW=16 Col=10 Bk=8 CS0 Row=17 CS1 Row=17 CS=2 Die BW=8 Size=4096MB
channel[2] BW=16 Col=10 Bk=8 CS0 Row=17 CS1 Row=17 CS=2 Die BW=8 Size=4096MB
channel[3] BW=16 Col=10 Bk=8 CS0 Row=17 CS1 Row=17 CS=2 Die BW=8 Size=4096MB
Manufacturer ID:0x1
CH0 RX Vref:29.7%, TX Vref:21.8%,21.8%
CH1 RX Vref:29.7%, TX Vref:22.8%,21.8%
CH2 RX Vref:28.9%, TX Vref:23.8%,22.8%
CH3 RX Vref:28.5%, TX Vref:21.8%,22.8%
change to F1: 528MHz
change to F2: 1068MHz
change to F3: 1560MHz
change to F0: 2112MHz
out
U-Boot SPL board init
U-Boot SPL 2017.09-g1b1171b7c71-220418 #fa (Apr 18 2024 - 11:12:44)
Trying to boot from MMC2
No misc partition
spl: partition error
Trying fit image at 0x4000 sector
## Verified-boot: 0
## Checking atf-1 0x00040000 ... sha256(b40ce543bb...) + OK
## Checking u-boot 0x00a00000 ... sha256(893641a38c...) + OK
## Checking fdt-1 0x00ae03a0 ... sha256(3e56afa830...) + OK
## Checking atf-2 0x000f0000 ... sha256(f202de33d9...) + OK
## Checking atf-3 0xff100000 ... sha256(af08976afd...) + OK
Jumping to U-Boot(0x00a00000) via ARM Trusted Firmware(0x00040000)
Total: 619.300/780.905 ms

INFO:    Preloader serial: 2
NOTICE:  BL31: v2.3():v2.3-616-gdc1125f48:finley.xiao
NOTICE:  BL31: Built : 16:41:17, Jul 13 2023
INFO:    spec: 0x1
INFO:    ext 32k is not valid
INFO:    ddr: stride-en 4CH
INFO:    GICv3 without legacy support detected.
INFO:    ARM GICv3 driver initialized in EL3
INFO:    valid_cpu_msk=0xff bcore0_rst = 0x0, bcore1_rst = 0x0
INFO:    system boots from cpu-hwid-0
INFO:    idle_st=0x21fff, pd_st=0x11fff9, repair_st=0xfff70001
ERROR:   dfs get fsp_params[0] error, 0xfead0004 != 0xfead0003
ERROR:   dfs get fsp_params[1] error, 0x1111 != 0xfead0003
ERROR:   dfs get fsp_params[2] error, 0x0 != 0xfead0003
ERROR:   dfs get fsp_params[3] error, 0x61251520 != 0xfead0003
ERROR:   loader&trust unmatch!!! Please update trust if need enable dmc
INFO:    BL31: Initialising Exception Handling Framework
INFO:    BL31: Initializing runtime services
WARNING: No OPTEE provided by BL2 boot loader, Booting device without OPTEE initialization. SMC`s destined for OPTEE will return SMC_UNK
ERROR:   Error initializing runtime service opteed_fast
INFO:    BL31: Preparing for EL3 exit to normal world
INFO:    Entry point address = 0xa00000
INFO:    SPSR = 0x3c9


U-Boot 2024.07 (Oct 25 2024 - 15:13:10 +0200)

Model: Radxa ROCK 5 Model B
DRAM:  16 GiB
Core:  353 devices, 32 uclasses, devicetree: separate
MMC:   mmc@fe2c0000: 1, mmc@fe2d0000: 2, mmc@fe2e0000: 0
Loading Environment from nowhere... OK
In:    serial@feb50000
Out:   serial@feb50000
Err:   serial@feb50000
Model: Radxa ROCK 5 Model B
Net:   No ethernet found.
Hit any key to stop autoboot:  0 
Scanning for bootflows in all bootdevs
Seq  Method       State   Uclass    Part  Name                      Filename
---  -----------  ------  --------  ----  ------------------------  ----------------
Scanning global bootmeth 'efi_mgr':
Card did not respond to voltage select! : -110
No EFI system partition
No EFI system partition
Failed to persist EFI variables
No EFI system partition
Failed to persist EFI variables
No EFI system partition
Failed to persist EFI variables
No EFI system partition
Failed to persist EFI variables
No EFI system partition
Failed to persist EFI variables
  0  efi_mgr      ready   (none)       0  <NULL>                    
** Booting bootflow '<NULL>' with efi_mgr
Loading Boot0000 'mmc 1' failed
Loading Boot0001 'mmc 0' failed
EFI boot manager: Cannot load any image
Boot failed (err=-14)
Scanning bootdev 'mmc@fe2c0000.bootdev':
  1  extlinux     ready   mmc          1  mmc@fe2c0000.bootdev.part /boot/extlinux/extlinux.conf
** Booting bootflow 'mmc@fe2c0000.bootdev.part_1' with extlinux
1:	FriendlyElec NanoPC T6 endpoint mode
Retrieving file: /boot/Image
append: root=/dev/mmcblk1p1 rw earlycon rootwait
Retrieving file: /boot/rk3588-nanopc-t6.dtb
Retrieving file: /boot/rk3588-nanopc-t6-pcie-ep.dtbo
## Flattened Device Tree blob at 12000000
   Booting using the fdt blob at 0x12000000
Working FDT set to 12000000
   Loading Device Tree to 00000000ecb68000, end 00000000ecb95fff ... OK
Working FDT set to ecb68000

Starting kernel ...

[    0.000000] Booting Linux on physical CPU 0x0000000000 [0x412fd050]
[    0.000000] Linux version 6.12.0-rc3 (rick@A13PC04R) (aarch64-none-linux-gnu-gcc (Arm GNU Toolchain 13.2.rel1 (Build arm-13.7)) 13.2.1 20231009, GNU ld (Arm GNU Toolchain 13.2.rel1 (Build arm-13.7)) 2.41.0.20231009) #6 SMP Wed Nov  6 13:20:41 CET 2024
```

**Comment by ste7anste7an** [comment]
Thanks for the update. You might want to change the device tree for the LTS version, as the USB ports do not get powered in the LTS version with the device three of the older model.
But everything is working quit well.

I have a question about an option rom, but I will make a new issue for that.

**Comment by ste7anste7an** [comment]
I close the issue for now.

**Comment by yigger** [comment]
Hi, I have the same problem.

My board is T6-LTS. I have flashed the official Ubuntu image into EMMC, and successfully flashed the image you provided in the release into the SD card. I also replaced extlinux.config.t6 with extlinux.conf, but it still boots from EMMC when it finally starts.

I sent the boot log to Gemini, and Gemini replied that
"According to the boot log, the current system is booted from eMMC, not from the SD card. Although SPL tried to boot from the SD card, it failed because no valid partition or image was found, and finally loaded U-Boot from eMMC."

So I want to ask if there is a problem with the release image? Or do I need to flash the image from scratch?

The boot log is below:

```
DDR 9fa84341ce typ 24/09/06-09:51:11,fwver: v1.18
ch0 ttot10
ch1 ttot10
ch2 ttot10
ch3 ttot10
ch0 ttot18
LPDDR4X, 2112MHz
channel[0] BW=16 Col=10 Bk=8 CS0 Row=17 CS1 Row=17 CS=2 Die BW=8 Size=4096MB
ch1 ttot18
channel[1] BW=16 Col=10 Bk=8 CS0 Row=17 CS1 Row=17 CS=2 Die BW=8 Size=4096MB
ch2 ttot18
channel[2] BW=16 Col=10 Bk=8 CS0 Row=17 CS1 Row=17 CS=2 Die BW=8 Size=4096MB
ch3 ttot18
channel[3] BW=16 Col=10 Bk=8 CS0 Row=17 CS1 Row=17 CS=2 Die BW=8 Size=4096MB
Manufacturer ID:0x1
DQS rds:h1,l0
CH0 RX Vref:30.5%, TX Vref:22.8%,22.8%
DQ rds:h1 l0 h1 l0 h1 l0 l0 h2, h2 l0 h2 h2 h4 h1 h1 h1

DQS rds:h1,l0
CH1 RX Vref:28.9%, TX Vref:22.8%,22.8%
DQ rds:h5 h1 h2 l0 h6 h2 h3 l0, h4 h1 h4 h3 l0 l0 h3 l0

DQS rds:l0,h1
CH2 RX Vref:28.9%, TX Vref:22.8%,21.8%
DQ rds:h4 h4 h1 h4 h2 h3 l0 h1, h2 h6 h1 h2 h3 h1 h1 h2

DQS rds:l1,h1
CH3 RX Vref:30.1%, TX Vref:22.8%,21.8%
DQ rds:h1 h1 h1 h1 h1 l0 h2 h4, h1 h2 h3 h4 h1 h2 l0 h5

stride=0x2, ddr_config=0x4
hash ch_mask0-1 0x20 0x40, bank_mask0-3 0xa00 0x1400 0x2800 0x0, rank_mask0 0x401000
change to F1: 528MHz
ch0 ttot10
ch1 ttot10
ch2 ttot10
ch3 ttot10
change to F2: 1068MHz
ch0 ttot14
ch1 ttot14
ch2 ttot14
ch3 ttot14
change to F3: 1560MHz
ch0 ttot16
ch1 ttot16
ch2 ttot16
ch3 ttot16
change to F0: 2112MHz
ch0 ttot18
ch1 ttot18
ch2 ttot18
ch3 ttot18
out
U-Boot SPL board init
U-Boot SPL 2017.09-g50821b3c8bd-241113 #fa (Jan 02 2025 - 16:53:42)
sfc cmd=0fH(6BH-x4)
Trying to boot from MMC2
No misc partition
spl: partition error
Trying fit image at 0x4000 sector
## Verified-boot: 0
## Checking atf-1 0x00040000 ... sha256(b40ce543bb...) + OK
## Checking u-boot 0x00a00000 ... sha256(893641a38c...) + OK
## Checking fdt-1 0x00ae03a0 ... sha256 Bad hash: af5c423c223127af9b0996137a17a02e404e109d1908ab484e406b0f67e2807e
 error!
Bad hash value for 'hash' hash node in 'fdt-1' image node
Trying fit image at 0x5000 sector
Not fit magic
Trying to boot from MMC1
SPL: A/B-slot: _a, successful: 0, tries-remain: 7
Trying fit image at 0x4000 sector
## Verified-boot: 0
## Checking atf-1 0x00040000 (gzip @0x00240000) ... sha256(f24113467f...) + sha256(7612223b82...) + OK
## Checking uboot 0x00200000 (gzip @0x00400000) ... sha256(627381f2ed...) + sha256(ba4b42d31c...) + OK
## Checking fdt 0x00349d50 ... sha256(9ff849197b...) + OK
## Checking atf-2 0xff100000 ... sha256(70505bb764...) + OK
## Checking atf-3 0x000f0000 ... sha256(b2af21b504...) + OK
## Checking optee 0x08400000 (gzip @0x08600000) ... sha256(05fd8c61c3...) + sha256(3e0a1f56ed...) + OK
Jumping to U-Boot(0x00200000) via ARM Trusted Firmware(0x00040000)
Total: 894.690/1096.875 ms

INFO:    Preloader serial: 2
NOTICE:  BL31: v2.3():v2.3-868-g040d2de11:derrick.huang, fwver: v1.48
NOTICE:  BL31: Built : 15:02:44, Dec 19 2024
INFO:    spec: 0x1
INFO:    code: 0x88
INFO:    ext 32k is not valid
INFO:    ddr: stride-en 4CH
INFO:    GICv3 without legacy support detected.
INFO:    ARM GICv3 driver initialized in EL3
INFO:    valid_cpu_msk=0xff bcore0_rst = 0x0, bcore1_rst = 0x0
INFO:    l3 cache partition cfg-0
INFO:    system boots from cpu-hwid-0
INFO:    disable memory repair
INFO:    idle_st=0x21fff, pd_st=0x11fff9, repair_st=0xfff70001
INFO:    dfs DDR fsp_params[0].freq_mhz= 2112MHz
INFO:    dfs DDR fsp_params[1].freq_mhz= 528MHz
INFO:    dfs DDR fsp_params[2].freq_mhz= 1068MHz
INFO:    dfs DDR fsp_params[3].freq_mhz= 1560MHz
INFO:    BL31: Initialising Exception Handling Framework
INFO:    BL31: Initializing runtime services
INFO:    BL31: Initializing BL32
I/TC:
I/TC: OP-TEE version: 3.13.0-898-g57604957e #hisping.lin (gcc version 10.2.1 20201103 (GNU Toolchain for the A-profile Architecture 10.2-2020.11 (arm-10.16))) #2 Fri Nov 15 16:04:27 CST 2024 aarch64, fwver: v1.19
I/TC: OP-TEE memory: TEEOS 0x200000 TA 0xc00000 SHM 0x200000
I/TC: Primary CPU initializing
I/TC: Primary CPU switching to normal world boot
INFO:    BL31: Preparing for EL3 exit to normal world
INFO:    Entry point address = 0x200000
INFO:    SPSR = 0x3c9


U-Boot 2017.09-g50821b3c8bd-241113 #fa (Jan 02 2025 - 16:53:42 +0800)

Board: FriendlyELEC RK3588 board
PreSerial: 2, raw, 0xfeb50000
DRAM:  16 GiB
Sysmem: init
Relocation Offset: eda14000
Relocation fdt: eb9faa20 - eb9fecd0
CR: M/C/I
Using default environment

optee api revision: 2.0
mmc@fe2c0000: 1, mmc@fe2e0000: 0
Bootdev(atags): mmc 0
MMC0: HS400 Enhanced Strobe, 200Mhz
PartType: EFI
TEEC: Waring: Could not find security partition
DM: v2
boot mode: None
RESC: 'resource', blk@0x0000a000
lib/avb/libavb_user/avb_ops_user.c: trusty_read_lock_state failed
Error determining whether device is unlocked.
Device is: UNLOCKED
DTB: rk3588-nanopi6-rev07.dtb (275186)
HASH(c): OK
ANDROID: fdt overlay OK
usb dr_mode not found
usb dr_mode not found
I2c0 speed: 100000Hz
vsel-gpios- not found!
en-gpios- not found!
vdd_cpu_big0_s0 800000 uV
vsel-gpios- not found!
en-gpios- not found!
vdd_cpu_big1_s0 800000 uV
I2c2 speed: 100000Hz
vsel-gpios- not found!
en-gpios- not found!
vdd_npu_s0 800000 uV
spi2: RK806: 2
ON=0x80, OFF=0x80
vdd_gpu_s0 750000 uV
vdd_cpu_lit_s0 init 800000 uV
vdd_log_s0 750000 uV
vdd_vdenc_s0 init 750000 uV
vdd_ddr_s0 850000 uV
get vp0 plane mask:0x7, primary id:0, cursor_plane:2, from dts
get vp1 plane mask:0x148, primary id:6, cursor_plane:3, from dts
get vp2 plane mask:0x280, primary id:7, cursor_plane:9, from dts
Could not find baseparameter partition
I2c6 speed: 200000Hz
Rockchip UBOOT DRM driver version: v1.0.1
vp0 have layer nr:3[0 1 2 ], primary plane: 0
vp1 have layer nr:3[3 6 8 ], primary plane: 6
vp2 have layer nr:2[7 9 ], primary plane: 7
vp3 have layer nr:0[], primary plane: 0
hdmi@fde80000 disconnected
CLK: (sync kernel. arm: enter 1008000 KHz, init 1008000 KHz, kernel 0N/A)
  b0pll 24000 KHz
  b1pll 24000 KHz
  lpll 24000 KHz
  v0pll 24000 KHz
  aupll 786431 KHz
  cpll 1500000 KHz
  gpll 1188000 KHz
  npll 850000 KHz
  ppll 1100000 KHz
  aclk_center_root 702000 KHz
  pclk_center_root 100000 KHz
  hclk_center_root 396000 KHz
  aclk_center_low_root 500000 KHz
  aclk_top_root 750000 KHz
  pclk_top_root 100000 KHz
  aclk_low_top_root 396000 KHz
vdd_usbc 12258 mV

Net:   No ethernet found.
Hit key to stop autoboot('CTRL+C'):  0
## Booting FIT Image FIT: No fit blob
FIT: No FIT image
ANDROID: reboot reason: "(none)"
Not AVB images, AVB skip
No valid android hdr
Android image load failed
Android boot failed, error -1.

## Booting Rockchip Format Image
fdt      @ 0x08300000 (0x00043592)
kernel   @ 0x00400000 (0x0221c808)
ramdisk  @ 0x0a200000 (0x00793fc0)
Fdt Ramdisk skip relocation
## Flattened Device Tree blob at 0x08300000
   Booting using the fdt blob at 0x08300000
   Using Device Tree in place at 0000000008300000, end 0000000008346591
## reserved-memory:
  cma: addr=10000000 size=8000000
  drm-logo@0: addr=edf00000 size=2ee000
  vendor-storage-rm@00000000: addr=ebcc0000 size=10000
  ramoops@110000: addr=110000 size=e0000
Adding bank: 0x00200000 - 0x08400000 (size: 0x08200000)
Adding bank: 0x09400000 - 0xf0000000 (size: 0xe6c00000)
Adding bank: 0x100000000 - 0x3fc000000 (size: 0x2fc000000)
Adding bank: 0x3fc500000 - 0x3fff00000 (size: 0x03a00000)
Adding bank: 0x4f0000000 - 0x500000000 (size: 0x10000000)
Total: 1544.977/1620.688 ms

Starting kernel ...

[    1.630212] Booting Linux on physical CPU 0x0000000000 [0x412fd050]
[    1.630237] Linux version 6.1.99 (fa@jensen) (aarch64-linux-gcc (ctng-1.26.0-129g-FA) 14.2.0, GNU ld (GNU Binutils) 2.43.1) #32 SMP Mon Jan 20 15:50:32 CST 2025
[    1.632894] random: crng init done
[    1.637321] Machine model: FriendlyElec NanoPC-T6 LTS
[    1.644300] earlycon: uart8250 at MMIO32 0x00000000feb50000 (options '')
[    1.647545] printk: bootconsole [uart8250] enabled
[    1.650127] OF: fdt: Reserved memory: failed to reserve memory for node 'drm-cubic-lut@0': base 0x0000000000000000, size 0 MiB
[    1.651210] OF: reserved mem: 0x0000000000110000..0x00000000001effff (896 KiB) map non-reusable ramoops@110000
[    1.652134] Reserved memory: created CMA memory pool at 0x0000000010000000, size 128 MiB

...

Ubuntu 24.04.1 LTS NanoPC-T6-LTS ttyFIQ0
```

**Comment by rick-heig** [comment]
Hello,
For some reason there seems to be problem with the SD card image, but I cannot replicate it.
May I ask how big the SD card you use is ?

I have tried with 32GB and 128GB SD cards and had no issues.

You could try to resizing the partitions following https://github.com/rick-heig/eNVMe/blob/main/doc/platform.md#4-resize-the-partition 

Check the partitions you have, you could use `fdisk` to create a new partition table : Here is the one on my SD :

```
Command (m for help): p

Disk /dev/sdg: 28.91 GiB, 31016878080 bytes, 60579840 sectors
Disk model: STORAGE DEVICE  
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: 4E41D4DF-436B-4635-8716-67CA9DC0AE82

Device     Start      End  Sectors  Size Type
/dev/sdg1  18716    51483    32768   16M Linux filesystem
/dev/sdg2  51484 60579806 60528323 28.9G Linux filesystem
```

If you use the same start and end for the first 16M partition, then the same start for the second partition, but let fdisk choose the size fit to your SD it should work.
(Use fdisk ‘d’ to delete all partitions, then create them with ’n’ etc.)

I don’t know if there is a need to redo the protective/hybrid MBR disklabel, but you can enter this mode with ‘M’ in fdisk and also redo the partition table there, here is mine :

```
Command (m for help): M

Entering protective/hybrid MBR disklabel.

Command (m for help): p
Disk /dev/sdg: 28.91 GiB, 31016878080 bytes, 60579840 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x00000000

Device     Boot Start    End Sectors  Size Id Type
/dev/sdg1  *    18716  51483   32768   16M  c W95 FAT32 (LBA)
/dev/sdg2       51484 575771  524288  256M 83 Linux
/dev/sdg3           1     33      33 16.5K ee GPT

Partition table entries are not in disk order.
```

Yes the partitions are not in order and not the correct size, IIRC this is only used to boot U-Boot, then when U-Boot launches Linux it uses the GPT table. (Not so sure anymore).

If you are unable to repartition your SD, you might try a 32GB SD or
You could also rebuild the SD from scratch, following all steps in https://github.com/rick-heig/eNVMe/blob/main/doc/platform.md this will definitely work but take some more time.

To save time you can do steps 1-4 which wouldn’t take so long.
Then remount the SD card, delete everything in the RootFS partition, mount the image file I provided as a loopback device so you can mount it on your computer and copy all files (with permissions) to the RootFS partition you just cleared by deleting all files.

Basically step 7 : https://github.com/rick-heig/eNVMe/blob/main/doc/platform.md#7-replace-the-old-rootfs-on-the-sd-with-the-new-one 
For the commands to mount the .img in loopback you can follow some commands of step 5 : https://github.com/rick-heig/eNVMe/blob/main/doc/platform.md#5-prepare-the-ubuntu-24041-rootfs 
(Mounting the compressed image may be possible, but you would have to search how, the steps above show how to mount an uncompressed image)

Anyways, the steps in https://github.com/rick-heig/eNVMe/blob/main/doc/platform.md allow you to build the image from scratch.

But I think it might just be an issue with the partition table, so doing https://github.com/rick-heig/eNVMe/blob/main/doc/platform.md#4-resize-the-partition on your flashed SD might be enough.

Let me know how it goes, and the size of your SD, maybe it is because I created the image with a 32 GB SD and flashing it onto smaller SDs might cause an issue with the partition table.

Best regards,
Rick

**Comment by yigger** [comment]
Thank you very much for your detailed reply.

I have solved it now.

My SD card is 128G. It may be due to the partition error that the boot was not installed correctly.

The following is my solution. Hope it can help guys who have the same problem as me:
```
1. Delete all partitions of the SD card, repartition, create ESP boot (it's important! ), and format the partition. (Use diskgenius partition tool)

2 Execute `xzcat sdcard_envme_ubuntu2404_241106.img.xz | sudo dd of=/dev/<your sd card> bs=1M && sudo sync`, wait for a long time, and successfully write to the SD card.

3. Modify the configuration file, mv extlinux.conf.t6 extlinx.conf

4. Reinsert the SD card and successfully enter the system.
```

Thank you again for your project and your enthusiastic reply.

By the way, I want to ask why envme-epf seems to only work on Linux? I connected Win10 via PCIE and executed the envme-epf command, but there seemed to be no response.

**Comment by rick-heig** [comment]
Glad to hear it is now working.

I have eNVMe working with a Windows 10 host.
Please take a look on the embedded board with `dmesg` for the messages related to nvme-epf, you can launch nvme-epf with --debug to see more messages.

Check if the PCIe link gets up, check if you receive commands from the host Windows machine, then in the host windows machine check the device manager if it sees the device or not and if it reports errors.

If there are still problems feel free to open another issue and share the logs.

Best regards,
Rick
